### ISSUE TITLE
question: how affines are applied to oblique datasets

### ISSUE THREAD

**Issue by oesteban** [afni/afni#116]
I'm applying an affine transform to a dataset where I artificially rotated the affine.

I get the habitual warning about the dataset>
```
++ 3dAllineate: AFNI version=AFNI_16.0.01 (Jan 27 2016) [64-bit]
++ Authored by: Zhark the Registrator
** AFNI converts NIFTI_datatype=2 (UINT8) in file someones_anatomy_rot.nii.gz to FLOAT32
     Warnings of this type will be muted for this session.
     Set AFNI_NIFTI_TYPE_WARN to YES to see them all, NO to see none.
*+ WARNING:   If you are performing spatial transformations on an oblique dset, 
  such as someones_anatomy_rot.nii.gz,
  or viewing/combining it with volumes of differing obliquity,
  you should consider running: 
     3dWarp -deoblique 
  on this and  other oblique datasets in the same session.
 See 3dWarp -help for details.
++ Oblique dataset:someones_anatomy_rot.nii.gz is 5.156967 degrees from plumb.
++ Oblique dataset:someones_anatomy_rot.nii.gz is 5.156967 degrees from plumb.
++ Source dataset: ./someones_anatomy_rot.nii.gz
++ Base dataset:   ./someones_anatomy_rot.nii.gz
++ You might want to use '-master' when using '-1D*_apply'
++ Loading datasets
++ master dataset for output = base
++ OpenMP thread count = 8
```

The question is: how does `3dAllineate` build the dicom 4x4 matrix for moving and reference image? Does it consider both? Looking at the results, the image's affines (ref and moving) seem to be first rotated to zero from plumb.

Bonus question: is there an executable in AFNI to which I input 4 homogeneous coordinates (x, y, z, 1) and returns them after applying the transform matrix (i.e. dot product with a transform given in DICOM 4x4 homogeneous coordinate system).

For a deeper context, check https://github.com/oesteban/BIDS-X5/blob/master/01%20-%20Preparing%20images.ipynb

**Comment by leej3** [comment]
Hi Oscar,

I picked @afni-dglen ‘s brains on this. Here’s a summary. We can go into more depth next week.

The definition of AFNI’s affine usage is described in the [vecwarp help](https://afni.nimh.nih.gov/pub/dist/doc/program_help/Vecwarp.html)  (at the -matvec entry):

3dAllineate is not a tool that deals with obliquity (the datasets deviation from plumb). Just like the AFNI viewer it discards the off-diagonal component of the affine so that it can consider the data in its un-interpolated form. It also uses backwards rather than forward notation for the affine (forwards/backwards notation are described in the vecwarp help and referred to elsewhere in afni help as “in2out”/”out2in” respectively).

> 3dAllineate build the dicom 4x4 matrix for moving and reference image?

It discards the off diagonal component of the images and computes the backward version of the affine  that aligns the two scans.

>  Does it consider both?

Yes but removes the obliquity of both.

> Looking at the results, the image's affines (ref and moving) seem to be first rotated to zero from plumb.

Correct

> is there an executable in AFNI to which I input 4 homogeneous coordinates (x, y, z, 1) and returns them after applying the transform matrix

Vecwarp will do this.

3dwarp will apply the affine transformation between an oblique dataset and another. So if you want to apply the matrix to an image as attempted in the notebook you linked…
```
cat_matvec  M.afni > M_afni_3by4.1D
3dWarp -matvec_in2out M_afni_3by4.1D -prefix moved-afni-3dwarp.nii.gz -newgrid 1.0  someones_anatomy_rot.nii.gz
```
The newgrid option is used to interpolate during the resampling. This would be especially important for anisotropic data. If you know the desired output grid (like one of the MNI templates) the gridset option is useful.

It is worth noting that if one wishes to align scans `align_epi_anat.py` is the tool to use. It will account for obliquity as it aligns the two input scans. Under the hood it uses 3dWarp and 3dAllineate.

# Other notes
A useful description of how afni deals with obliquity is [here](https://sscc.nimh.nih.gov/sscc/dglen/Obliquity).

AFNI refers to dicom order as RAI (I believe nibabel would say LPS+).

**Comment by oesteban** [comment]
Thanks @leej3 that is very useful.

At this point, I'm trying to write transforms that would behave as if AFNI had computed them. The idea is to start with an affine matrix _T_ in physical coordinates (i.e., there's no concept of image orientation because it is just a function that maps coordinates in one space _x_ to another _y_).

so _y = T(x)_.

When the image axes are aligned with the cardinal axes, I've found that prepending and appending a reorientation to LPS matrix (LPS in nibabel's argot) does the trick:

_M'  = LPS . T . LPS_, where _M'_ is the affine that I write to disk and then check with ``3dAllineate``

The problem arises with oblique datasets because of the implicit realignment of the axis that AFNI introduces. So my intent here is to find that _M'_ such that the rotation matrices of reference and moving images are concatenated with _T_.

**Comment by afni-dglen** [comment]
3dWarp computes something like this in the course of warping one dataset to another dataset, where one or both may be oblique with the -card2oblique option. To do this, it computes an xyz->xyz transformation. You can see this with the -verb option and in 3dWarp.c's source code. The transformation is basically a concatenation of the undoing (inverting here) of the closest cardinal equivalent matrix (ijk->xyz_card) and the ijk->xyz_real with the reverse order for a second dataset. The overall concatenated matrix is then defined for an xyz_card->xyz_card. Half of this occurs for just "deobliquing" a dataset, that is transforming a dataset onto the closest cardinal grid. 

I'm not sure which of these you want,  but I think these will be good leads for you on obliquity. The align_epi_anat.py script may be another useful place to look for examples of how AFNI uses the oblique information. Generally, at least for now, AFNI uses the oblique xyz coordinates in only a few places as in that script.

**Comment by oesteban** [comment]
> 3dWarp computes something like this in the course of warping one dataset to another dataset, where one or both may be oblique with the -card2oblique option.

Does this mean that 3dWarp and 3dAllineate will behave differently when provided with the same affine matrix in the case that source and/or base image are "oblique"?. If so, what would be the _canonical_ behavior for AFNI: 3dWarp's or 3dAllineate's?

> You can see this with the -verb option and in 3dWarp.c's source code

Thanks! This is very helpful, will look at this carefully.

> I'm not sure which of these you want, but I think these will be good leads for you on obliquity. The align_epi_anat.py script may be another useful place to look for examples of how AFNI uses the oblique information. Generally, at least for now, AFNI uses the oblique xyz coordinates in only a few places as in that script.

The end goal of this is to be able to read, say, an FSL affine matrix and write out the same transform in a way that AFNI tools will apply the transform correctly. Also the other way around, and including other software tools that produce/consume affines (ITK, FreeSurfer, SPM, etc.)

**Comment by afni-dglen** [comment]
The computation of the obliquity transformation is handled only by 3dWarp.  Both can apply transformation matrices in similar ways. 3dWarp can apply in the in2out or out2in direction. 3dAllineate applies transformations only using the matrix from the out2in direction. 3dAllineate is much more commonly used, but the obliquity transformation is computed using 3dWarp and then concatenated with any other affine transformations, like alignment of EPI to anat and anat to  template. 3dNwarpApply extends that with nonlinear warps and uses the same affine transformations that 3dAllineate uses.

Some other relevant information:

cat_matvec is the principal tool for affine transformation concatenation. It can read the AFNI header, 1D text files with transformations on one line or multiple lines, invert transformations and pull out the rigid equivalent transformation. Nonlinear warps in combination with affine transformations are computed and applied with 3dNwarpApply instead.

whereami has a little known tool for concatenating transformations among spaces and finds the shortest path to transform among spaces, concatenating affine transformations where it is possible. The options -calc_chain and -show_chain find the chain of transformations between source and destination spaces. The spaces and transformations may be defined in a SessionAtlases.niml file, in a similar format as the AFNI_atlas_spaces.niml file in the binary directory. It can handle a variety of transformations, but it doesn't yet include nonlinear warping. It's sometimes a useful tool for combining transformations and getting the order and direction right.

**Comment by oesteban** [comment]
Hi thanks a lot for all the feedback. I'm pretty positive I just need one more thing to get my adaptor rolling.

I think that [this function](https://github.com/afni/afni/blob/83f330f20bab3491cc7645a4cb31f49b54f97883/src/3dAllineate.c#L3753) or something very similar is used to calculate the affine matrix without rotation w.r.t. the cardinal axes.

How does afni calculate the center of coordinates (i.e., rows 0-2 of column 3 of the 4x4 matrix)?
In other words, how `dax->xxorg` and so forth are set (e.g., here https://github.com/afni/afni/blob/b6a9f7a21c1f3231ff09efbd861f8975ad48e525/src/thd_matdaxes.c#L252).

**Comment by afni-dglen** [comment]
xxorg, yyorg and zzorg are  the coordinates for the center of the first voxel in the dataset (ijk=0,0,0)

**Comment by oesteban** [comment]
Yes, but they don't seem to match those of the origin of the original file (before `3dWarp -deoblique`) nor the coordinates that result from applying the (inverse) rotation matrix to the oblique affine - I'm not 100% sure of the second one, and that's why I'm asking.

**Comment by oesteban** [comment]
Sorry for the double posting. Let me explain it in another way.

So I have an image _I_ of matrix size 57x67x56, which affine matrix _A_, copied on both s/q-forms and s/q-form code "scanner", is:
```
[[ -2.75   0.     0.    76.  ]
 [  0.    -2.75   0.    90.5 ]
 [  0.     0.     2.75 -91.  ]
 [  0.     0.     0.     1.  ]]
```
Then I generate a pure rotation matrix (i.e., no translation) _R_.
Finally, I generate an "oblique" image _I'_, just changing its s/q-form matrix to contain _R . A_. That gives the affine matrix _A'_:
```
[[-2.74999725e+00  2.74999817e-03  2.74999954e-03  7.58184241e+01]
 [-2.98603540e-03 -2.73886840e+00 -2.47165887e-01  9.83951362e+01]
 [ 2.49170222e-03 -2.47168626e-01  2.73886865e+00 -8.25664202e+01]
 [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  1.00000000e+00]]
```
Next, I run `3dWarp -deoblique -prefix deob.nii.gz i_prime.nii.gz`, and the affine of `deob.nii.gz` is _B_:
```
[[ -2.75        -0.          -0.          77.35987854]
 [ -0.          -2.75        -0.          98.75680542]
 [  0.           0.           2.75       -99.2093277 ]
 [  0.           0.           0.           1.        ]]
```
Therefore, there is an offset that breaks _A_ = _B_

If I try to check the _R_ estimated with `3dWarp -deoblique` as _Rafni_ = _B_ . _A^-1_ I can see that all the rotation elements are the same as in _R_, but there is a translation:
```
[[ 9.99999000e-01  1.08583106e-03 -9.06073535e-04  1.35987854e+00]
 [-9.99999333e-04  9.95952145e-01  8.98795002e-02  8.25680542e+00]
 [ 9.99999833e-04 -8.98785043e-02  9.95952235e-01 -8.20932770e+00]
 [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  1.00000000e+00]]
```
Where does that translation come from? How can I reconstruct it?

**Comment by afni-dglen** [comment]
I'm not sure that bit of code you linked actually gets used for any of this. The 3dWarp -deoblique command uses the Compute_Deoblique_Transformation function in 3dWarp.c. You can see the transformation applied with 3dWarp -disp_obl_xform -deoblique ... or use the "-verb" option. The applied transformation is also stored in the AFNI header or AFNI extension within the NIFTI header. Does that help?

**Comment by oesteban** [comment]
Yes, that helped a bit. So I went ahead to try to find out the info in the NIFTI header. To recap a bit, I'm checking the NIfTI1 extension of the dataset written out by `3dWarp -deoblique`:

```xml
<?xml version='1.0' ?>
<AFNI_attributes
  self_idcode="XYZ_ueMSwlle6F8B1j0JMBCMQg"
  NIfTI_nums="58,72,62,1,1,16"
  ni_form="ni_group" >
<AFNI_atr
  ni_type="String"
  ni_dimen="1"
  atr_name="HISTORY_NOTE" >
 "[oesteban@dendrite: Mon Oct  7 15:38:14 2019] 3dWarp -deoblique -prefix deoblps.nii.gz someones_anatomy_lpsrot.nii.gz"
</AFNI_atr>

<AFNI_atr
  ni_type="float"
  ni_dimen="12"
  atr_name="WARPDRIVE_MATVEC_FOR_000000" >
 0.999999
 0.001085831
 0.0009060735
 0.1815765
 -0.0009999992
 0.9959521
 -0.0898795
 -7.895141
 -0.0009999998
 0.0898785
 0.9959521
 8.433576
</AFNI_atr>

<AFNI_atr
  ni_type="float"
  ni_dimen="12"
  atr_name="WARPDRIVE_MATVEC_INV_000000" >
 0.999999
 -0.0009999993
 -0.0009999998
 -0.1810379
 0.001085831
 0.9959522
 0.08987851
 7.104988
 0.0009060736
 -0.08987951
 0.9959523
 -9.109215
</AFNI_atr>

<AFNI_atr
  ni_type="String"
  ni_dimen="1"
  atr_name="TYPESTRING" >
 "3DIM_HEAD_ANAT"
</AFNI_atr>

<AFNI_atr
  ni_type="String"
  ni_dimen="1"
  atr_name="IDCODE_STRING" >
 "XYZ_ueMSwlle6F8B1j0JMBCMQg"
</AFNI_atr>

<AFNI_atr
  ni_type="String"
  ni_dimen="1"
  atr_name="IDCODE_DATE" >
 "Mon Oct  7 15:38:14 2019"
</AFNI_atr>

<AFNI_atr
  ni_type="int"
  ni_dimen="8"
  atr_name="SCENE_DATA" >
 0
 11
 0
 -999
 -999
 -999
 -999
 -999
</AFNI_atr>

<AFNI_atr
  ni_type="String"
  ni_dimen="1"
  atr_name="LABEL_1" >
 "zyxt"
</AFNI_atr>

<AFNI_atr
  ni_type="String"
  ni_dimen="1"
  atr_name="LABEL_2" >
 "zyxt"
</AFNI_atr>

<AFNI_atr
  ni_type="String"
  ni_dimen="1"
  atr_name="DATASET_NAME" >
 "zyxt"
</AFNI_atr>

<AFNI_atr
  ni_type="int"
  ni_dimen="3"
  atr_name="ORIENT_SPECIFIC" >
 0
 3
 4
</AFNI_atr>

<AFNI_atr
  ni_type="float"
  ni_dimen="3"
  atr_name="ORIGIN" >
 -77.35988
 -98.75681
 -99.20933
</AFNI_atr>

<AFNI_atr
  ni_type="float"
  ni_dimen="3"
  atr_name="DELTA" >
 2.75
 2.75
 2.75
</AFNI_atr>

<AFNI_atr
  ni_type="float"
  ni_dimen="12"
  atr_name="IJK_TO_DICOM" >
 2.75
 0
 0
 -77.35988
 0
 2.75
 0
 -98.75681
 0
 0
 2.75
 -99.20933
</AFNI_atr>

<AFNI_atr
  ni_type="float"
  ni_dimen="12"
  atr_name="IJK_TO_DICOM_REAL" >
 2.75
 0
 0
 -77.35988
 0
 2.75
 0
 -98.75681
 0
 0
 2.75
 -99.20933
</AFNI_atr>

<AFNI_atr
  ni_type="float"
  ni_dimen="2"
  atr_name="BRICK_STATS" >
 0
 95.66844
</AFNI_atr>

<AFNI_atr
  ni_type="int"
  ni_dimen="8"
  atr_name="DATASET_RANK" >
 3
 1
 0
 0
 0
 0
 0
 0
</AFNI_atr>

<AFNI_atr
  ni_type="int"
  ni_dimen="5"
  atr_name="DATASET_DIMENSIONS" >
 58
 72
 62
 0
 0
</AFNI_atr>

<AFNI_atr
  ni_type="int"
  ni_dimen="1"
  atr_name="BRICK_TYPES" >
 3
</AFNI_atr>

<AFNI_atr
  ni_type="float"
  ni_dimen="1"
  atr_name="BRICK_FLOAT_FACS" >
 0
</AFNI_atr>

<AFNI_atr
  ni_type="String"
  ni_dimen="1"
  atr_name="TEMPLATE_SPACE" >
 "ORIG"
</AFNI_atr>

<AFNI_atr
  ni_type="int"
  ni_dimen="1"
  atr_name="INT_CMAP" >
 0
</AFNI_atr>

</AFNI_attributes>
```

I'm guessing we are interested in the `WARPDRIVE_MATVEC_FOR_000000` attribute. As we can see, the rotation elements match very well what I find out from the headers (and it matches the inverse of my initial _R_ matrix, so that looks good too):

AFNI: 
```
 0.999999  0.001085831  0.0009060735 0.1815765
 -0.0009999992  0.9959521  -0.0898795  -7.895141
 -0.0009999998  0.0898785  0.9959521  8.433576
```
numpy/nibabel
```
9.99999000e-01  1.08583106e-03 -9.06073535e-04  1.35987854e+00
-9.99999333e-04  9.95952145e-01  8.98795002e-02  8.25680542e+00
9.99999833e-04 -8.98785043e-02  9.95952235e-01 -8.20932770e+00
```

Despite that annoying axis flip that I'll need to figure out myself, it is clear that the `WARPDRIVE_MATVEC_FOR_000000` transform is not a pure rotation, and three translation components are still present. The question is, therefore, where these translation components come from? If we are only "deobliquing" a dataset, wouldn't a pure-rotation matrix suffice?

Thanks very much.

**Comment by oesteban** [comment]
Okay, the offset I'm seeing is the offset between the origin of coordinates of the oblique dataset and its rotation back to plumb via the deoblique affine. Thanks @afni-dglen for the help!

**Comment by oesteban** [comment]
Sorry to reopen with a follow-up question (esp. two years in)

I just understood why I wasn't getting the exact translation values on the WARPDRIVE matrices, when compared to those in the nifti header extension. Something I wasn't expecting is that the oblique image is resliced on a new grid (which is always isotropic) with the deobliqued matrix:

```
Using minimum spacing of 2.750000 mm for new grid spacing
```

so, my question now is how the new grid size is calculated? I've reverse-engineered that the center of the data array stays at the same world coordinate where it was in the original oblique dataset. However, I'm seeing that the matrix size is grown and I can't figure out from the code where the new grid is calculated, which I'm guessing should be around here:

https://github.com/afni/afni/blob/75766463758e5806d938c8dd3bdcd4d56ab5a485/src/3dWarp.c#L641

**Comment by oesteban** [comment]
Gentle nudge for @afni-dglen :)

Thanks in advance for your patience :+1:

**Comment by afni-dglen** [comment]
Hi Oscar,
Sorry for the delay. For the simpler case of deobliquing, 3dWarp undoes the cardinal matrix and applies the ijk_to_dicom_real matrix (basically the same as the sform matrix with reversed signs for the first two rows).

T = Tc^-1 * Tr

where Tc is the ijk_to_dicom (cardinal) matrix, and Tr is the ijk_to_dicom_real matrix.

https://github.com/afni/afni/blob/75766463758e5806d938c8dd3bdcd4d56ab5a485/src/3dWarp.c#L838


The product is applied to the xyz coordinates based on a cardinal matrix to put every xyz coordinate into the "real" coordinates within a bounding box that contains the transformed box. The minimum voxel size is used by default in 3dWarp for isotropic voxels in the output dataset. One can override that in 3dWarp with -gridset to set a master grid instead or the -newgrid option for a specified isotropic voxel size.

**Comment by oesteban** [comment]
> The product is applied to the xyz coordinates based on a cardinal matrix to put every xyz coordinate into the "real" coordinates within a bounding box that contains the transformed box.

Thanks for taking the time to attend to my resurrection of the issue :)

Okay, if I understand correctly, then the 8 corners of the original grid are "deobliqued" (projected through the rotation matrix that brings the physical axes to be aligned with the data array axes). Then the extreme coordinates are calculated (i.e. to calculate the XYZ coordinates of the bounding box), and finally the new grid's size is adjusted so that it contains the XYZ bounding box. Is that correct?

**Comment by afni-dglen** [comment]
The language is a bit different, but that all seems right. The field of view is set by default by warping the bounding corners. The bounding box may also be set by the -gridset option to specify a dataset.

https://github.com/afni/afni/blob/75766463758e5806d938c8dd3bdcd4d56ab5a485/src/mri_warp3D.c#L1161
